FX.cpp: Update "colorful" for flexibility in # of colors from palette

zanhecht opened the pull request:
Instead of hard-coding the "Colorful" effect to have an override to use 5 colors for C9_2, add a slider to manually specify the number of colors to select from the palette, and if that slider is set to 0, automatically detect the number of colors in the palette (up to 16). This allows it to be used with custom palettes with more than 4 colors.

Also, change description of saturation slider to indicate its actual function (switching between "easter" colors, default colors, and the three color slots). I wasn't even aware that "saturation" could allow using the color slots until I looked at the source code.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Colorful mode now supports up to 16 configurable colors (including an auto/0 option), with palette sampling and a non-repeating color selection.
  * Rendering was updated to apply colors in fixed blocks across the strip while safely wrapping at strip end; low-intensity behavior remains pastel/default.

* **Documentation**
  * Updated the mode’s on-screen metadata to show **Color scheme** and **# of colors**, including the new default configuration.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

zanhecht committed:
FX.cpp: Update colorful for flexibility in # of colors from palette

Instead of hard-coding an override to use 5 colors for C9_2, add a function to automatically detect the number of colors in the palette (up to 16), so it can be used with custom palettes of more than 4 colors. Add a slider to manually specify number of colors, and change description of saturation slider to indicate its actual function (switching between "easter" colors, default colors, and the three color slots).

zanhecht committed:
FX.cpp: Cleanup code in "Colorful" for detecting number of colors

Change the logic so the 0th step isn't coded separately, only access SEGPALETTE once per loop instead of three times.

DedeHai commented:
I like the addition, there are a few issues:
- instead of changing the name of "saturation" to a lengthy string, such a description should be in KB not UI
- dito for the c3 slider
- added sliders should if at all possible not use a default value, it breaks existing presets as they default to center position (16 in this case)

actual saturation could be implemented using the `adjust_color()` function (lighten to desaturate) but that would also break existing presets that use a palette, so I am not sure its a good idea to do it, although it would allow to use pastel colors on all palettes.

zanhecht committed:
FX.cpp: Colorful: Add fallback if numColors == 0, shorten slider names

zanhecht committed:
FX.cpp: mode_colorful: Add missing value to RGBW

zanhecht committed:
FX.cpp: mode_colorful: avoid out-of-bounds writes

When numColors > SEGLEN or SEGLEN % numColors != 0, i + j can exceed SEGLEN-1. Guard the inner loop to prevent OOB writes.

DedeHai commented:
>Would the details just be included in https://kno.wled.ge/features/effects/ in the "Description" column for the effect as a whole, or is there some other place where more detailed information on effects is presented?

that is where I would suggest to add info for FX yes. You can update the description and do a PR with a comment that it should only be merged after 0.16 is released as KB should document released versions. Also include that these new functionality is only available >0.16

> That seems like a bug in presets that should be fixed (have them default to the default value if there is one and nothing is specified for that slider), but I understand that that's often easier said than done. 

all slider and option values are included in a preset, if the slider is not used, it saves the default value, which is just the slider center and "false" for unused options. So not something that is fixable unfortunately for existing presets.

>Any ideas of another way to implement this without changing the current behavior of grabbing 4 evenly-spaced colors? I'd prefer not to do something unintuitive like have it interpret 16 as 4.

rename the slider to "colors" so its not tied to a slider number but just arbitrary and map the slider such that the center is 4. I would also do the "auto" at max value and keep a minimum of 2 at a zero input value, which seems more intuitive since auto does 16 colors on most palettes.

zanhecht commented on wled00/FX.cpp:
```diff
@@ -1012,28 +1012,40 @@ static const char _data_FX_MODE_CHASE_RAINBOW_WHITE[] PROGMEM = "Rainbow Runner@
-  unsigned numColors = 4; //3, 4, or 5
-  uint32_t cols[9]{0x00FF0000,0x00EEBB00,0x0000EE00,0x000077CC};
+  unsigned numColors = (SEGMENT.custom3 > 16) ? 16 : SEGMENT.custom3; //defaults to 4, max of 16
+  uint32_t cols[31]{0x00FF0000,0x00EEBB00,0x0000EE00,0x000077CC};
```
The original version of this effect did the same thing. The pattern is repeated twice in the array, and the program reads 0-15, 1-16, 2-17, 3-18, etc. to create the scrolling effect. I suppose this was simpler than writing code that would wrap around to the beginning of the array in the middle of the sequence.
